velox/functions/prestosql/GreatestLeast.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <functions/Macros.h>
+#include <cmath>
+#include <type_traits>
+#include "velox/common/base/Exceptions.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/type/Type.h"
+
+namespace facebook::velox::functions {
+
+template <typename TExec, typename TInput, bool isLeast>
+struct ExtremeValueFunction;
+
+template <typename TExec, typename TInput>
+using LeastFunction = ExtremeValueFunction<TExec, TInput, true>;
+
+template <typename TExec, typename TInput>
+using GreatestFunction = ExtremeValueFunction<TExec, TInput, false>;
+
+/**
+ * This class implements two functions:
+ *
+ * greatest(value1, value2, ..., valueN) → [same as input]
+ * Returns the largest of the provided values.
+ *
+ * least(value1, value2, ..., valueN) → [same as input]
+ * Returns the smallest of the provided values.
+ **/
+template <typename TExec, typename TInput, bool isLeast>
+struct ExtremeValueFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExec);
+
+  // For double, presto should throw error if input is Nan
+  template <typename T>
+  void checkNan(const T& value) const {
+    if constexpr (std::is_same_v<T, TypeTraits<TypeKind::DOUBLE>::NativeType>) {
```

Comment:
This behavior is quite surprising. One would expect this functionality to apply to both DOUBLE and REAL types, not only to DOUBLE. It might be helpful to open an issue in PrestoDB repo to ask whether this logic is intentional.

Replace `TypeTraits<TypeKind::DOUBLE>::NativeType` with `double` for readability.

Reply:
Issue created 
https://github.com/prestodb/presto/issues/22391

Reply:
@mbasmanova Based on their reply, could we safely remove the NaN checks?

Reply:
@Real-Chen-Happy Yes, we can remove NaN checks, but then we need to make sure that NaN is considered larger than any non-NaN value.